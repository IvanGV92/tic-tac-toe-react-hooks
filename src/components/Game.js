import React, { useState } from 'react';
import { calculateWinner } from '../helpers';
import Board from './Board';
import './styles/Game.css';


const Game = () => {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [stepNumber, setStepNumber] = useState(0);
  const [xIsNext, setXisNext] = useState(true);
  const winner = calculateWinner(history[stepNumber]);

  const handleClick = i => {
    const timeInHistory = history.slice(0, stepNumber + 1);
    const current = timeInHistory[stepNumber];
    const squares = [...current];

    if (winner || squares[i]) return;

    squares[i] = xIsNext ? 'X' : 'O';
    setHistory([...timeInHistory, squares]);
    setStepNumber(timeInHistory.length);
    setXisNext(!xIsNext);
  }

  const rewindMoves = () =>{
   const rewindCount = (stepNumber > 0) ? stepNumber-1 : 0;
    jumpTo(rewindCount);
  }

  const reset =()=>{
    setHistory([Array(9).fill(null)]);
    setStepNumber(0);
    setXisNext(true);

  }

  const jumpTo = step => {
    setStepNumber(step);
    setXisNext(step % 2 === 0)
  };




  return (
    <>
      <Board squares={history[stepNumber]} onClick={handleClick} />
      <div className="game-div">
        <p className="info-message">{winner ? 'Winner: ' + winner : (stepNumber === 9) ? "It's a tie!" : 'Next Player: ' + (xIsNext ? 'X' : 'O')}</p>
        <button className="action-button" onClick={rewindMoves}>Rewind</button>
        <button className="action-button" onClick={reset}>Reset</button>

      </div>
    </>
  )
}

export default Game;