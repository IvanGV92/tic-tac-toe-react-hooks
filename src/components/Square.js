import React from 'react';

const style ={
  background: '#122343',
  border: '2px solid white',
  fontSize: '30px',
  fontWeight: '80',
  cursor:'pointer',
  outline: 'none',
  color:'white'

}

const Square = ( { value, onClick }) => (
  <button style={style} onClick={onClick}>
    {value}
  </button>
);

export default Square;