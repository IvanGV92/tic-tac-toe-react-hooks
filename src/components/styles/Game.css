.game-div{
  width: 200px;
  margin: 20px auto;
  font-size: 21px;
  font-weight: 400;
  font-family: 'Helvetica Neue', Helvetica, Arial, Roboto, sans-serif;
}


.info-message{
  margin-left:16%;
}

.action-button{
  background-color: #122343;
  border: 15px solid #fff;
  color:#fff;
  font-weight: 500;
  font-size: 1rem;
}

