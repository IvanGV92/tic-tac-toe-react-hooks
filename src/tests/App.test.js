
import React from 'react';
import '@testing-library/jest-dom';

import { shallow, mount} from 'enzyme';

import Square from '../components/Square';
import Board from '../components/Board';
import Game from '../components/Game';



describe("App",()=>{
  describe("Testing Board",()=>{
    it("should render a Square correctly",()=>{
      const value = null;
      const onClick = jest.fn();

      const wrapper = shallow(
      <Square
        value={value}
        onClick={onClick}
      />);

      expect(wrapper).toMatchSnapshot();

    });


    it("should render the Board correctly", () => {
      const squares = [Array(9).fill(null)][0];
      const onClick = jest.fn();

      const wrapper = shallow(
        <Board
          squares={squares}
          onClick={onClick}
        />);

      expect(wrapper).toMatchSnapshot();

    });

    it("should render next player label correctly", () => {


      const textLabel = "Next Player: X";
      const wrapper = shallow(
        <Game/>);


      const nextPlayerLabel = wrapper.find('p').text();

      expect(nextPlayerLabel).toBe(textLabel);

    });
    it("should render rewind button", () => {


      const buttonLabel = "Rewind";
      const wrapper = shallow(
        <Game />);


      const RewindButtonLabel = wrapper.find('button').at(0).props().children;

      expect(RewindButtonLabel).toEqual(buttonLabel);


    });

    it("should render reset button", () => {


      const buttonLabel = "Reset";
      const wrapper = shallow( <Game />);


      const ResetButtonLabel = wrapper.find('button').at(1).props().children;



      expect(ResetButtonLabel).toEqual(buttonLabel);


    });

  });

  describe("Testing Game", () => {


    describe('Reset', () => {
      const wrapperGame = shallow(<Game />);




      it('resets game', () => {

        const squares = [Array(9).fill(null)][0];
        const onClick = jest.fn();
        const wrapperBoard = shallow(
          <Board
            squares={squares}
            onClick={onClick}
          />);

        const btn = wrapperGame.find('button').at(1);

        btn.simulate('click');

        const board = wrapperGame.find('Board').html();



        expect(board).toBe(wrapperBoard.html());

      });
    });



  });





});