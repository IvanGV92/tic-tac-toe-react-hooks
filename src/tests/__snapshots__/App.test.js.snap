// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Testing Board should render a Square correctly 1`] = `
<button
  onClick={[MockFunction]}
  style={
    Object {
      "background": "#122343",
      "border": "2px solid white",
      "color": "white",
      "cursor": "pointer",
      "fontSize": "30px",
      "fontWeight": "80",
      "outline": "none",
    }
  }
/>
`;

exports[`App Testing Board should render the Board correctly 1`] = `
<div
  style={
    Object {
      "border": "4px solid #122343",
      "borderRadius": "10px",
      "display": "grid",
      "gridTemplate": "repeat(3, 1fr) / repeat(3, 1fr)",
      "height": "250px",
      "margin": "20px auto",
      "width": "250px",
    }
  }
>
  <Square
    key="0"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="1"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="2"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="3"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="4"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="5"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="6"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="7"
    onClick={[Function]}
    value={null}
  />
  <Square
    key="8"
    onClick={[Function]}
    value={null}
  />
</div>
`;
